### DISCUSSION TITLE
`ImTools` as an SDK repo for `ImSwitch`

### DISCUSSION THREAD

**Discussion by jacopoabramo** [ImSwitch/ImSwitch#210] [Ideas]
@kasasxav @fbalzarotti @nornil @beniroquai @Kubiczek36 

During the Lund hackaton organized by @nornil I created a repository called [ImTools](https://github.com/ImSwitch/ImTools). The objective I have is tied to the plugin system implementation: move the core parts of ImSwitch (i.e. all the base classes that provide the foundation for implementing new widget-controller-manager trios) to a "toolkit" repository that can be used to create new plugins without the need to download ImSwitch or access its source code. You can think of it as both a refactor of current existing ImSwitch code as an enhancement. Furthermore, I'm taking this opportunity to rewrite the documentation from scratch (at least for developers that want to create new trios) and also enhance code quality via type hinting. It would also greatly help to make sure that the "core" of ImSwitch is more formally unit-tested.

With this toolkit, what I hope to achieve is that:

- developers have more control over testing their plugin;
- they're less scared of dealing with a huge codebase;
- core objects are more easily tested and verified;
- ImSwitch boot-up time improves significantly:
  - rather than doing a huge list of imports in all the ImSwitch modules, only the plugins selected in a configuration are imported

I'm currently stalling a bit because I'm in the critical point of choosing weather or not to port into this repository the `CommunicationChannel` and the `ModuleCommunicationChannel` as well. The reason is because all controllers have a reference to these objects as they can interface respectively with the rest of the `imcontrol` module (which is usually our target) and to other existing ImSwitch modules currently active in the session.

I'm also considering the option to move the communication channels to another separate repository, i.e. `ImComm` or something along these lines.

The problem I encountered with ImSwitch so far is that there are so many nested things that address testing that it is hard to separate. By chopping up parts of ImSwitch that could be more easily managed and tested, the problem should be addressed.

I'm not aiming for making any breaking changes with the current ImSwitch status, but like I said is more a process of tidying up the current situation.

The challenge I see is that this repository currently holds the core classes of the `imcontrol` and `imcommon`module, and there might be a desire to move both the `imscripting` and the `imreconstruct` modules core parts in this repo (or alternatively different repos) as well; hence this would mean having different repos to manage. But at the same time a more "divide-et-impera" approach at this stage might significantly help when it comes down to maintainance. Maybe it'll take more time but it'll be less confusing.

So the questions are:
- would you agree with this course of action in the prospect of making developer's life more easy when it comes down to develop new widgets? 
- what possible challenges do you foresee by moving the communication channels to a separate repository (either this `ImTools` or another one)?

**Comment by palec87** [comment]
https://github.com/ImSwitch/ImTools 404 error. Isn't it private perhaps?

**Reply by jacopoabramo** [reply]
Yes you're right, I forgot that I initially set it to private. I'll open it now

**Reply by jacopoabramo** [reply]
Try now

**Comment by xcasas** [comment]
I like a lot the idea to separate ImSwitch into multiple repos. I'm not sure about the separation of the communication channels themselves, but maybe because I miss something. I agree with having the modules separated from each other, especially imcontrol since it's the main one and more used.

From my current research I have some possible ideas:

- You could have one project for the ImSwitch api: the general classes needed to develop ImSwitch plugins. This would be the project that you have to import to create a widget, controller, and manager. For example: [WidgetController & its factory](https://github.com/ImSwitch/ImSwitch/blob/master/imswitch/imcommon/controller/basecontrollers.py), [Widget & Widget Factory](https://github.com/ImSwitch/ImSwitch/blob/master/imswitch/imcontrol/view/widgets/basewidgets.py), [Manager & MultiManager](https://github.com/ImSwitch/ImSwitch/blob/master/imswitch/imcontrol/model/managers/MultiManager.py#L8).

- Another project for classes that can be useful when developing. This could be base widgets like [NapariHybridWidget](https://github.com/ImSwitch/ImSwitch/blob/master/imswitch/imcontrol/view/widgets/basewidgets.py), controllers like LiveUpdatedController, SuperScanController (basically the [basecontrollers](https://github.com/ImSwitch/ImSwitch/blob/master/imswitch/imcontrol/controller/basecontrollers.py)), and the basic managers like DetectorManager, LasersManager, etc.
The point of this would be to have classes that the developer can extend, so this project is only imported when needed. For example: LaserFromVendorManager, would import the LaserManager from this project (and Manager from the first project).

- Then the imswitch core (well, imcontrol). This could be communication channels but not only. Also: MasterController, SetupInfo, etc. And even the basic widgets like ViewWidget (although this I'm not so sure about).

- Finally, another with all available widgets, controllers & managers. This could also include the plugin system that you mention.

What do you think?
Actually this separation would only be for the imcontrol module. The others I would just keep them as they are and in separate repos.

**Comment by jacopoabramo** [comment]
The reason why I'm thinking of moving the communication channels to ImTools is because, for example, the [base controller class](https://github.com/ImSwitch/ImTools/blob/87d9cc2ff33bf0b0fa77b3c81b7c01c94f3b67e0/imtools/imcontrol/controller/basecontrollers.py#L3) that takes in itialization a reference to the CommunicationChannel class, and same goes for the [widget controller](https://github.com/ImSwitch/ImTools/blob/87d9cc2ff33bf0b0fa77b3c81b7c01c94f3b67e0/imtools/imcommon/controller/basecontrollers.py#L14) class. I think I still have to wrap my mind around the various levels of dependency that ImSwitch has internally.

At any rate the final goal is to have a codebase that respects a more strict approach towards type hinting. This is something that for me - personally - helps a lot when exploring the code, and it could also be helpful for developers when they want to do their implementation.

I already started porting the setup info classes into this repository, and I will probably modify it in order to being able to dynamically load different setup info dataclasses that can be defined by the users in order to add further functionalities via the configuration file.

Still the problem is that by this reasoning the MasterController should also go in this repository. At any rate we were thinking of changing a bit the recording part of ImSwitch with a more solid acquisition engine based off pycromanager that should be able to control simultaneously both Micro-Manager and ImSwitch devices, and the MasterController would be the center piece of that (not now, but we hope to get there soon once we have a fully tested prototype).

EDIT: of course if there are smarter ways to enforce type hinting in the class contructors I'm open to suggestions, but with the experience I witnessed so far with developers dealing with creating their widget stack, I feel this would be really helpful

**Reply by xcasas** [reply]
In my opinion you could have it all in the ImSwitch api repo. Anything that the developer needs can be there. But you don't need to have the whole communication channel implementation, just an interface with the functions.
Now I'm actually working with Java so you would have to check how to do it in Python. But you could have a very simple CommunicationChannel in the API so that the developer can use it when developing Widgets, and then in the ImSwitch core you could have the implementation of the CommunicationChannel which implements the CommunicationChannel interface. Do you follow me?

This could be done with all classes that the developer needs. This way you keep all the ImSwitch main functionality in the core, and all the interfaces in the API, so the developer imports the API and can develop applications/plugins.  Then you have this second project with various implementations that can be useful (my bullet point 2).

EDIT: And you could potentially change and have other interfaces that hide away the communication channel. The main point of the channel was to have a set of signals to interconnect components but you could choose to do it in other ways. This way the core would always stay the same, and the interfaces would facilitate the interaction with the developer.
